frontend/ui/menusorter.lua

```diff
@@ -0,0 +1,166 @@
+            menu_table[order_id] = item_table[order_id]
+            menu_table[order_id].id = order_id
+            for order_number,order_number_id in ipairs(order_item) do
+                -- this is a submenu, mark it for later
+                if order[order_number_id] then
+                    table.insert(sub_menus, order_number_id)
+                    tmp_menu_table[order_number] = {
+                        id = order_number_id,
+                    }
+                -- regular, just insert a menu action
+                else
+                    if order_number_id == "----------------------------" then
+                        -- it's a separator
+                        tmp_menu_table[order_number] = self.separator
+                    elseif item_table[order_number_id] ~= nil then
+                        item_table[order_number_id].id = order_number_id
+                        tmp_menu_table[order_number] = item_table[order_number_id]
+                        -- remove reference from item_table so it won't show up as orphaned
+                        item_table[order_number_id] = nil
+                    end
+                end
+            end
+            -- compress menus
+            -- if menu_items were missing we might have a table with gaps
+            -- but ipairs doesn't like that and quits when it hits nil
+            local i = 1
+            local new_index = 1
+            while i <= table.maxn(tmp_menu_table) do
+                local v = tmp_menu_table[i]
+                if v then
+                    if v.id == "----------------------------" then
+                        new_index = new_index - 1
+                        menu_table[order_id][new_index].separator = true
+                    else
+                        -- fix the index
+                        menu_table[order_id][new_index] = tmp_menu_table[i]
+                    end
+
+                    new_index = new_index + 1
+                end
+                i = i + 1
+            end
+        else
+            DEBUG("menu id not found:", order_id)
+        end
+    end
+
+    -- now do the submenus
+    for i,sub_menu in ipairs(sub_menus) do
+        local sub_menu_position = self:findById(menu_table["KOMenu:menu_buttons"], sub_menu) or nil
+        if sub_menu_position and sub_menu_position.id then
+            sub_menu_position.sub_item_table = menu_table[sub_menu]
+            -- remove reference from top level output
+            menu_table[sub_menu] = nil
+            -- remove reference from input so it won't show up as orphaned
+            item_table[sub_menu] = nil
+        end
+    end
+    -- @TODO avoid this extra mini-loop
+    -- cleanup, top-level items shouldn't have sub_item_table
+    for i,top_menu in ipairs(menu_table["KOMenu:menu_buttons"]) do
+        menu_table["KOMenu:menu_buttons"][i] = menu_table["KOMenu:menu_buttons"][i].sub_item_table
+    end
+
+    -- handle disabled
+    if order.KOMenu__disabled then
```

Comment:
This key's naming convention is inconsistent with other magic keys like `KOMenu:menu_buttons`. We should either name it as `order['KOMenu:disabled']` or rename `KOMenu:menu_buttons` to `KOMenu__menu_buttons`.